### ISSUE TITLE
RFC proposal - Add validation rule for special characters

### ISSUE THREAD

**Issue by bravo-kernel** [cakephp/cakephp#7418]
Figuring out how to validate if special characters are present in a password took me more time than I had expected and I can imagine more people need this. Before creating a PR for a new ``specialChars`` rule in the Validation class please let me know if this is even desired. If not... just close. 

The following regex basically uses a negate pattern to treat everything not matching `a-z`, `A-Z` and `0-9` as a special character.

```php
$regex = '/([^\x30-\x39\x61-\x7A\x41-\x5A])/';
```
Could be added with a count parameter.

**Comment by markstory** [comment]
Defining which characters are 'special' can be tricky. It looks like you are defining it as non-alphanumeric characters. Perhaps the rule name can reflect that vs. special?

**Comment by bravo-kernel** [comment]
You are correct, this is not limited to special characters so a different rule name would be appropriate. Maybe something like ``nonAlphaNumerics($count)``, suggestions welcome. 

Below the final regex as intended, in readable format and using lookahead for the counter. In this case the string would need to contain two non-alphanumeric characters to pass validation. 

```php
$regex = '/(?=([^a-zA-Z0-9]){2})/';
```

Please let me know if this could pass so I can create a PR.

**Comment by markstory** [comment]
```php
$validator->add('password', 'complexity', ['rule' => ['containsNonAlphaNumeric', 2]]);
```
Reads pretty well I think

**Comment by bravo-kernel** [comment]
Sure does Mr. Story, I will start work on the PR

**Comment by HavokInspiration** [comment]
This issue was addressed with #7457.
Changing milestone back to 3.1.1 where the feature was added.
